File: apps/sandbox-go/executor/processExecutor.go
package executor

import (
	"bytes"
	"os/exec"
	"sandbox-go/utils"
)

type ProcessExecutor struct{}

func NewProcessExecutor() *ProcessExecutor{
	return  &ProcessExecutor{}
}

func (e *ProcessExecutor) Exec(req utils.ExecRequest) utils.ExecResponse{
	cmd := exec.Command(req.Command, req.Args...)

	if req.Cwd != ""{
		cmd.Dir = req.Cwd
	}

	env := make([]string, 0, len(req.Env))
	for k,v := range req.Env {
		env = append(env, k+"="+v)
	}
	if len(env) > 0 {
		cmd.Env = append(cmd.Env, env...)
	}

	var stdout bytes.Buffer
	var stderr bytes.Buffer

	cmd.Stdout = &stdout
	cmd.Stderr = &stderr

	err := cmd.Run()

	exitCode := 0

	if err != nil{
		if exitErr, ok := err.(*exec.ExitError); ok {
			exitCode = exitErr.ExitCode()
		}else{
			return utils.ExecResponse{
				Stdout: "",
				Stderr: "",
				ExitCode: -1,
				Error: err.Error(),
			}
		}
	}
	return utils.ExecResponse{
		Stdout:   stdout.String(),
		Stderr:   stderr.String(),
		ExitCode: exitCode,
	}
}

File: apps/sandbox-ts/executor/processExecutor.ts
import { ExecRequest, ExecResponse } from "../utils/types";
import { ChildProcess, spawn } from "child_process";

export interface StartedProcess {
    process: ChildProcess;
}

export class ProcessExecutor {

    start(req: ExecRequest): StartedProcess {
        const child = spawn(req.command, req.args, {
            env: { ...process.env, ...req.env },
            cwd: req.cwd,
            stdio: ["ignore", "pipe", "pipe"]
        });

        return {
            process: child,
        };
    }

    async exec(req: ExecRequest): Promise<ExecResponse> {
        return new Promise((resolve) => {
            const child = spawn(req.command, req.args, {
                env: { ...process.env, ...req.env },
                cwd: req.cwd
            })

            let stdout = "";
            let stderr = "";

            child.stdout.on("data", (data) => {
                stdout += data.toString()
            })

            child.stderr.on("data", (data) => {
                stderr += data.toString()
            })

            child.on("error", (err) => {
                resolve({
                    stdout: "",
                    stderr: "",
                    exitCode: -1,
                    error: err.message
                })
            })

            child.on("close", (code) => {
                resolve({
                    stdout,
                    stderr,
                    exitCode: code ?? -1
                })
            })
        })
    }
}